central/telemetry/phonehome/interceptor.go

```diff
@@ -60,6 +62,19 @@ func getRequestDetails(ctx context.Context, err error, info *grpc.UnaryServerInf
 		method = "unknown"
 		code = -1
 	}
+
+	if throughGrpcGateway {
+		// By default, all permanent HTTP headers are added grpcgateway- prefix:
+		// https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/mux.go#L106-L114
+		// User-Agent header is occupied with internal grpc-go value:
+		// https://github.com/grpc/grpc-go/blob/0238b6e1cec37b55820b461d3d30652c54efe2c4/clientconn.go#L211-L215
+		userAgentKey := fmt.Sprintf("%suser-agent", runtime.MetadataPrefix)
+		userAgent = strings.Join(ri.Metadata.Get(userAgentKey), ", ")
+	}
+	// Fallback to HTTP
+	if !throughGrpcGateway || userAgent == "" {
+		userAgent = strings.Join(ri.Metadata.Get("User-Agent"), ", ")
+	}
```

Comment:
May we just try to find the `grpcgatewayuser-agent` key and use it if found or fallback to `User-Agent` otherwise?

Reply:
I think the code here has clearer intent: if `grpc-gateway` is used, use `grpc-gateway` header; if it does not work **or** we are not on gRPC protocol, use HTTP header. It would be unexpected to see `grpcgateway-user-agent` in the HTTP interceptor, and I would think we should ignore such header in such context.

Reply:
I think there is too much code for this intent: you even had to add more parameters to two functions and compute the normal User-Agent twice.
Instead, you could move this grpc related check (i.e. try to read the `grpcgatewayuser-agent` header, and you don't need any bool parameter) to the beginning of the function, and put the existing User-Agent retrieval under `else`.

Reply:
➕